src/unix/signal.c

```diff
@@ -454,13 +454,32 @@ static void uv__signal_event(uv_loop_t* loop,
 
     for (i = 0; i < end; i += sizeof(uv__signal_msg_t)) {
       msg = (uv__signal_msg_t*) (buf + i);
-      handle = msg->handle;
-
-      if (msg->signum == handle->signum) {
-        assert(!(handle->flags & UV_HANDLE_CLOSING));
-        handle->signal_cb(handle, handle->signum);
+      /* Passing pointers via pipes is not portable (it is not possible e.g. on
+       * CHERI-enabled architectures, where reading a pointer from a pipe will
+       * yield a non-dereferenceable value as it clears a hidden validity
+       * bit). To find the handler in a portable way, we look up it up inside
+       * uv__signal_tree using the address of the handler.
+       */
+      handle = NULL;
+      for (handle = uv__signal_first_handle(msg->signum);
+           handle != NULL && handle->signum == msg->signum;
+           handle = RB_NEXT(uv__signal_tree_s, &uv__signal_tree, handle)) {
```

Comment:
Iterating over the r/b-tree is unsafe unless `uv__signal_lock()` is called first.

Acquiring the lock and walking the tree adds quite a bit of overhead and I don't think it's reasonable to make users on other platforms pay for that, so there would have to be a CHERI-specific code path.

Signal handlers block on the lock so you can't hold it for too long. You will definitely have to drop it before the `handle->signal_cb` callback because callbacks can run for an indefinite amount of time.